src/cart/crt_rpc.c

```diff
@@ -1216,6 +1216,9 @@ crt_req_hg_addr_lookup(struct crt_rpc_priv *rpc_priv)
 	if (rc != 0) {
 		crt_context_req_untrack(rpc_priv);
 		crt_rpc_complete(rpc_priv, rc);
+
+		/* Do not propagate error further as we've completed the rpc */
+		rc = DER_SUCCESS;
```

Comment:
Thanks, Alex.

Hmm, if it just return SUCCESS, then crt_req_send_internal() will return SUCCESS in this case, and it seems there are a few extra error handling depending on the rc of the function?

Reply:
most of the error handling is supposed to be done by crt_rpc_complete() along with req_untrack() when needed. Ideally we should not have crt_req_send() (which calls crt_req_send_internal()) return an error back to the user, except in rare circumstances when things like parameters are wrong or some allocation fails

Most of the error reporting is supposed to be done via completion callback (which is invoked by crt_rpc_complete()).

We already have similar logic during actual sending of rpc if HG_Forward() returns error. in such case we mark rpc as unreachable and eventually execute following code, but dont propagate the error further up.
 842         case RPC_STATE_FWD_UNREACH:
 843                 RPC_ERROR(rpc_priv,
 844                           "failed due to group %s, rank %d, tgt_uri %s can't reach the target\n",
 845                           grp_priv->gp_pub.cg_grpid,
 846                           tgt_ep->ep_rank,
 847                           rpc_priv->crp_tgt_uri);
 848                 crt_context_req_untrack(rpc_priv);
 849                 crt_rpc_complete(rpc_priv, -DER_UNREACH);

Reply:
Hmm, do you think  
  if (rc != 0)
                rpc_priv->crp_state = RPC_STATE_INITED;
at the end of crt_req_send_internal()  should be moved into crt_rpc_complete()

Reply:
At that point it should not matter. Once we invoked completion callback the rpc will be finalized, and no next steps/stages will be invoked

Reply:
ah, ok